Introduce BrotliDecoder

slandelle opened the pull request:
Motivation:

Netty lacks client side support for decompressing Brotli compressed response bodies.

Modification:

* Introduce optional dependency to brotli4j by @hyperxpro. It will be up to the user to provide the brotli4j libraries for the target platform in the classpath. brotli4j is currently available for Linux, OSX and Windows, all for x86 only.
* Introduce BrotliDecoder in codec module
* Plug it onto `HttpContentDecompressor` for HTTP/1 and `DelegatingDecompressorFrameListener` for HTTP/2
* Add test in `HttpContentDecoderTest`

Result:

Netty now supports decompressing Brotli compressed response bodies.

slandelle commented:
cc @hyperxpro

slandelle commented:
@normanmaurer @hyperxpro The PR fails randomly on Linux (first build failed on Java 8, second one failed on Java 11 while Java 8 was successful) with the following error:

> 2021-01-25T22:09:01.2229566Z 22:09:01.212 [main] DEBUG i.n.handler.codec.compression.Brotli - Failed to load brotli4j; Brotli support will be unavailable.
2021-01-25T22:09:01.2232104Z java.lang.UnsatisfiedLinkError: /tmp/com_aayushatharva_brotli4j_1234063592518/libbrotli.so: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/com_aayushatharva_brotli4j_1234063592518/libbrotli.so)

Do you guys have any idea? Is it an issue with GH Actions where hosts are inconsistent? It there any way to compile the native lib differntly?

hyperxpro commented:
Looks like a dependency error. @normanmaurer Help :D

normanmaurer commented:
@slandelle @hyperxpro we use Centos 6 on our CI setup to ensure the maximum compatibility. Your native artefacts were compiled on something which has GLIBC_2.29 which is a lot newer to what cantos 6 offers. For maximum compatibility you should ensure you compile it on a system with a very old GLIBC version.

hyperxpro commented:
I used GitHub Actions for building the native library.

normanmaurer commented on codec-http/src/test/resources/jquery-3.5.1.min.js:
```diff
@@ -0,0 +1,2 @@
+/*! jQuery v3.5.1 | (c) JS Foundation and other contributors | jquery.org/license */
```
@slandelle we will need to update our NOTICE file and add the license of jquery to the license folder.

slandelle replied:
This is just a sample. I'm going so switch to a different file, eg Netty's HTML home page so we don't have such dependency and don't end up raising security bots alarms when there's a security issue with a jQuery version.

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/Brotli.java:
```diff
@@ -0,0 +1,64 @@
+    ClassNotFoundException cnfe = null;
+
+    try {
+      Class.forName("com.aayushatharva.brotli4j.Brotli4jLoader", false,
```
we also use 4 spaces... This needs to be fixed everywhere

slandelle replied:
Will fix.

But frankly, Netty formatting is very complicated to handle, in particular for contributors who can't remember and switch IDE settings for each project they contribute to.
IMHO, Netty should implement something like [spotless](https://github.com/diffplug/spotless) to take care of automatic formatting directly plugged onto maven.

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,135 @@
+  @Override
+  protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+    if (in.readableBytes() == 0) {
+      out.add(in);
```
this doesn't look correct... Remove this line. We should just return

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    if (decoder == null) {
+      decoder = new DecoderJNI.Wrapper(8 * 1024);
```
I think this should happen in `handlerAdded(...)`

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  public void channelInactive(ChannelHandlerContext ctx) throws Exception {
```
missing `@Override`

normanmaurer commented:
@hyperxpro yeah this explains it as they use ubuntu 18 (as minimum). You will need to use something like docker to compile with centos6. Thats what we do

normanmaurer commented on pom.xml:
```diff
@@ -580,6 +580,26 @@
+        <groupId>com.aayushatharva.brotli4j</groupId>
+        <artifactId>native-windows-x86_64</artifactId>
+        <version>1.3.0</version>
+      </dependency>
```
I think these should be optional as its an optional dependency really

slandelle replied:
I'm not fond of having `<optional>` in `<dependencyManagement>`. IMHO, that's something to be defined on each module that use the deps. Eg, codec-http has this dep with `test` scope, no need for `optional`.
I see that defining `optional` in <dependencyManagement>` is the current coding style so I'll use it.

slandelle replied:
Actually, not defining `<optional>` in parent's `<dependencyManagement>` but defining it in the `child` that requires this dep is consistent with how things are done for all the other codecs, eg https://github.com/netty/netty/blob/4.1/pom.xml#L578-L582 vs https://github.com/netty/netty/blob/4.1/codec/pom.xml#L81-L85.

What would you rather have?

normanmaurer replied:
sounds good

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  private State decompress(ByteBuf input, List<Object> output) {
+    while (true) {
```
please use `for (;;)` to be consistent with our coding style

franz1981 commented:
Just a comment: there is no chance to get a JNI-free pure Java impl?

hyperxpro commented:
> Just a comment: there is no chance to get a JNI-free pure Java impl?

Nope, not yet.

slandelle commented:
> Just a comment: there is no chance to get a JNI-free pure Java impl?

The java implementation has been lagging behind for a very long time. [Last release was in May 2017](https://search.maven.org/artifact/org.brotli/dec).

franz1981 commented:
@slandelle 
> The java implementation has been lagging behind for a very long time. Last release was in May 2017.

Recently I see the good trend to get some JNI code turned into Java one, even while accessing native resources and I like it, for many reasons.
I'm looking on https://github.com/google/brotli/tree/master/java/org/brotli if is pure-Java or not...

slandelle commented:
@franz1981 franz1981 

Assuming similar levels of performance and maintenance, I agree I'd rather have a pure Java implementation than a JNI one.
Sadly, so far:

* the Brotli team hasn't been actively maintaining the pure Java implementation, on contrary to the C one.
* I've never seen a benchmark comparing the performance of the Java based implementation. I would expect le latter to be implemented mostly for the sake of the client side on Android, where performance is not really an issue (low throughput per client instance).

slandelle commented:
@normanmaurer I've just addressed your comments, thanks!
@hyperxpro The ball is now in your court :) Please chime in once you've fixed the glibc issue (sorry, I can't help with that).

hyperxpro commented:
@normanmaurer CentOS 6 has officially reached EOL. You sure we need to compile Linux binary on CentOS 6?

normanmaurer commented:
@hyperxpro while it is EOL I still think it is a good idea for max compatibility

slandelle commented:
> @hyperxpro while it is EOL I still think it is a good idea for max compatibility

After some digging, I second @normanmaurer here: the glibc version on the latest Amazon Linux 2 is 2.26, so we definitely require 2.29.

hyperxpro commented:
What's the GLIBC version I need to compile this with?

slandelle commented:
@chrisvest 
WDYM? This is a new and optional feature.

slandelle commented:
> What's the GLIBC version I need to compile this with?

From [distrowatch](https://distrowatch.com/table.php?distribution=centos), CentOS 6 has glibc 2.12 while CentOS 7 has glibc 2.17.

normanmaurer commented:
Yes just use cents 6 which uses glib 2.12... OpenJDK compiles against and old GLIBC for the same reasons as well.

hyperxpro commented:
@normanmaurer Would you mind helping me build? My docker experience is almost `null`.

 :-|

normanmaurer commented:
@hyperxpro sorry I am super busy atm so I can't help a lot here. But basically this should give you some idea:

https://github.com/netty/netty/blob/4.1/docker/Dockerfile.centos6
https://github.com/netty/netty/blob/4.1/docker/docker-compose.yaml
https://github.com/netty/netty/blob/4.1/docker/docker-compose.centos-6.18.yaml

hyperxpro commented:
I give up Docker. WSL 2 is making me cry. @slandelle Can you help in building the binary of Linux?

idelpivnitskiy commented on codec-http/pom.xml:
```diff
@@ -67,6 +67,30 @@
+    <dependency>
+      <groupId>commons-io</groupId>
+      <artifactId>commons-io</artifactId>
+    </dependency>
```
This should have `test` scope

slandelle replied:
Done is parent's dependencyManagement block: https://github.com/netty/netty/pull/10960/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R763-R768

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/Brotli.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public static boolean isAvailable() {
```
Consider adding javadoc for all public methods and the class

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,136 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public final class BrotliDecoder extends ByteToMessageDecoder {
```
Consider adding javadoc

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    private void destroy() {
+        if (destroyed.compareAndSet(false, true)) {
```
Does it have to be protected if it always invoked from a single event loop thread?

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,136 @@
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        decoder = new DecoderJNI.Wrapper(8 * 1024);
```
IIUC, the decoder is initialized here to allow reusing it. However, the internal state is tiny, it should be ok to create a new instace instead of removing and adding the existing one. Consider initializing decoder in the constructor, then the `decoder` can be `final`.

normanmaurer replied:
just a note if we do this in the constructor we will also need to destroy it in a finalizer as you never know if the user ends up adding the handler to the pipeline

slandelle replied:
@normanmaurer That was my original reasoning for making this lazy. However, is this really a concern as this Decoder  is only created when the response has a br content-encoding? see https://github.com/netty/netty/pull/10960/files#diff-2e33eaddd20ffe54e6c4e266fc9cf90ac5f8715ec82a7bcef5d45dbcfdee4996R72

idelpivnitskiy replied:
Not a concert, just a suggestion. Works for me either way.
Another questions: should we allow users (who wants to use `BrotliDecoder` directly, not for HTTP) to configure `inputBufferSize`?

slandelle replied:
> Another questions: should we allow users (who wants to use BrotliDecoder directly, not for HTTP) to configure inputBufferSize?

good idea, will do

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,136 @@
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (!in.isReadable()) {
+            return;
```
Should we also check that the decoder is not `destroyed`? Other decoders (Zlib, Bzip2) skip all new bytes after they are done.

slandelle replied:
OK

idelpivnitskiy commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:
```diff
@@ -159,6 +184,70 @@ public void testResponseDecompression() {
+    }
+
+    @Test
+    public void testResponseChunksBrotliDecompression() {
```
Thanks for adding tests for HTTP. Can you please also add a `BrotliDecoderTest` as a subclass of `AbstractDecoderTest`? It will run the common tests we have for other decoders and you can add more brotli-specific tests there if necessary.

slandelle replied:
done

idelpivnitskiy reviewed (changes requested):
Thank you @slandelle!

hyperxpro commented:
@slandelle Go ahead with this. I don't have plans to downgrade the GLIBC version. `Brotli4jLoader#isAvailable` call should be enough.

slandelle commented:
@hyperxpro I'll try to have someone at Gatling take a look next week at the Docker/GH Action configuration.

As is:

* this PR can't be merged as it doesn't pass Netty's CI
* brotli4j doesn't work on current major distributions, including Amazon Linux 2. From my (Gatling) PoV, that's a major blocker.

hyperxpro commented:
I agree with your point. I also saw some issues with RHEL 8.X.

slandelle commented:
Some update on this: we (at Gatling) can't manage to compile the original Brotli code on CentOS6. We plan on contributing fixes to brotli4j so that Linux binaries work on widespread modern distributions (eg current Amazon Linux) and keep on integrating on our side.

As for contributing upstream to Netty, until Netty revisits its CentOS compatibility requirement, the only solution I see is to disable Brotli related tests on linux-x86_64.

Thoughts?

hyperxpro commented:
@slandelle Agree with your points. Also, please do the PR for Brotli4j.

slandelle commented:
@hyperxpro Yeah, we're working on this. We're still working on the CentOS docker image as it's still quite large.

slandelle commented:
Just updated PR with brotli4j 1.4.0, which is compatible with CentOS 6.

hyperxpro commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,163 @@
+
+    private final int inputBufferSize;
+    private DecoderJNI.Wrapper decoder;
+    private boolean destroyed = false;
```
No need to explicitly initialize it to `false` because that's the default value.

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,164 @@
+            if (state == State.DONE) {
+                destroy();
+            } else if (state == State.ERROR) {
+                throw new DecompressionException("Brotli stream corrupted");
```
Can brotli recover from errors? If not, consider invoking `destroy()` method before throwing an exception.

slandelle replied:
done

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,164 @@
+    private ByteBuf pull(ByteBufAllocator alloc) {
+        ByteBuffer nativeBuffer = decoder.pull();
+        ByteBuf copy = alloc.buffer(nativeBuffer.remaining());
+        copy.writeBytes(nativeBuffer);
```
Please, provide an inline comment for future developers with explanation why the content of returned `nativeBuffer` have to be copied, and can not be wrapped as-is or as `nativeBuffer.duplicate()`.

slandelle replied:
done

idelpivnitskiy reviewed (approved):
LGTM after the last minor comments,
Thank you @slandelle!

slandelle commented on codec/src/main/java/io/netty/handler/codec/compression/Brotli.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Throws when brotli support is missinf from the classpath or is unavailable on this platform
```
done

slandelle commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,164 @@
+        this(8 * 1024);
+    }
+
+    public BrotliDecoder(int inputBufferSize) {
```
done

slandelle commented on codec/src/test/java/io/netty/handler/codec/compression/BrotliDecoderTest.java:
```diff
@@ -0,0 +1,45 @@
+
+public class BrotliDecoderTest extends AbstractDecoderTest {
+
+  protected BrotliDecoderTest() throws Exception {
```
done

hyperxpro commented:
@slandelle Do you have plans for Brotli Encoder or I should add it to my list?

slandelle commented on codec/src/test/java/io/netty/handler/codec/compression/BrotliDecoderTest.java:
```diff
@@ -0,0 +1,45 @@
+public class BrotliDecoderTest extends AbstractDecoderTest {
+
+  protected BrotliDecoderTest() throws Exception {
+    Brotli4jLoader.ensureAvailability();
```
Brotli's c++ code requires a moderately recent c++ compiler, way more recent to what's available on CentOS6 which is Netty's Linux x86 CI target.
Backporting this compiler to CentOS6 is pretty complicated and we're still struggling. We'll keep on trying for a while but this is a backgroung task for us, not a priority.
If we disable the tests when brotli is not available in the platform, Netty's CI won't run those tests for x86_64. I'm not sure if this is fine.

Generally speaking, I really hope Netty will revisit this CentOS6 requirement in a future version.
Also, I'm wondering if Netty's CI is sufficient to prove that all JNI components work on distributions more modern than CentOS6.

hyperxpro replied:
> > Do you have plans for Brotli Encoder or I should add it to my list?
> 
> @hyperxpro no, no plans, sorry. In Gatling, we're mainly focused on the client side, so it's very unlikely we'll ever work on the encoder ourselves. And we're still working on trying to make the binary work on CentOS6.

Alright. I will do PR for the encoder after we're done fixing the binary issue and merger of this PR.

idelpivnitskiy replied:
//cc @normanmaurer @chrisvest

chrisvest replied:
We want to bump the OS and glibc versions For Netty 5, but I don't think we can do that in a patch release. That said, the aach build is already using CentOS 7. We could add a x86_64 build on CentOS as well.

normanmaurer replied:
yeah I think it would be fine to run the tests etc as long as we "release" out native stuff with the old GLBC

normanmaurer replied:
so I think we basically should remove this call and do something like:


```java
@BeforeClass
public static void assumeBrotli() {
    Assume.assumeTrue(Brotli.isAvailable());
}
```

slandelle replied:
Running tests on CentOS7 would be great. I think my coworker who's trying to compile brotli for CentOS6 is losing his hair.

normanmaurer replied:
lmao... If you can show me the error I might be able to give some hints. I did fight this battle a few times with BoringSSL etc ;)

slandelle replied:
Great news, my coworker was able to build brotli in a compatible fashion. We're going to send a new PR for brotli4j, and then update this one.

normanmaurer replied:
nice!

hyperxpro replied:
@slandelle Good news. Hope your coworker is fine. :p

slandelle commented:
> Do you have plans for Brotli Encoder or I should add it to my list?

@hyperxpro no, no plans, sorry. In Gatling, we're mainly focused on the client side, so it's very unlikely we'll ever work on the encoder ourselves. And we're still working on trying to make the binary work on CentOS6.

hyperxpro commented:
> > Do you have plans for Brotli Encoder or I should add it to my list?
> 
> @hyperxpro no, no plans, sorry. In Gatling, we're mainly focused on the client side, so it's very unlikely we'll ever work on the encoder ourselves. And we're still working on trying to make the binary work on CentOS6.

Alright. I will do PR for the encoder after we're done fixing the binary issue and merger of this PR.

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -48,16 +48,24 @@
 
+    /**
+     * Creates a new BrotliDecoder
+     * @param inputBufferSize desired size of the input buffer in kB
```
It's in bytes, not kB

slandelle replied:
facepalm

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:
```diff
@@ -159,6 +185,81 @@ public void testResponseDecompression() {
+
+    @Test
+    public void testResponseChunksBrotliDecompression() {
+        assertTrue("Brotli should be available on this platform", Brotli.isAvailable());
```
I think it would be fine to use an `assumeTrue(...)` here so the build can complete on older platforms as well.

slandelle replied:
OK

hyperxpro reviewed (commented):
Is sample data random or it contains real information like email?

hyperxpro commented:
```
Error:  COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
Error:  /home/runner/work/netty/netty/codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:[225,33] error: unreported exception Throwable; must be caught or declared to be thrown
[INFO] 1 error
[INFO] -------------------------------------------------------------
```

hyperxpro commented:
ByteBuf leak ;/ @slandelle

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:
```diff
@@ -159,6 +186,83 @@ public void testResponseDecompression() {
+    @Test
+    public void testResponseBrotliDecompression() throws Throwable {
+//        Assume.assumeTrue(Brotli.isAvailable());
+        Brotli.ensureAvailability();
```
This needs to get removed and the above line needs to be "uncommented"

slandelle replied:
I beg to differ.
If we manage to make Brotli4j work on all OS tested in Netty's CI, we shouldn't use Assume.
Otherwise, if an update of Brotli4j breaks (eg 1.4.0 was missing the native lib for MacOS), we won't be able to detect it as the test will be automatically silenced.

normanmaurer replied:
thats actually a very good point.

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/HttpContentDecoderTest.java:
```diff
@@ -159,6 +186,83 @@ public void testResponseDecompression() {
+    @Test
+    public void testResponseChunksBrotliDecompression() throws Throwable {
+//        Assume.assumeTrue("Brotli support is not available on this platform", Brotli.isAvailable());
+        Brotli.ensureAvailability();
```
same comment as above

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,173 @@
+     * @param inputBufferSize desired size of the input buffer in bytes
+     */
+    public BrotliDecoder(int inputBufferSize) {
+        this.inputBufferSize = inputBufferSize;
```
I think we need to verify this is positive or I am missing something ?

slandelle replied:
done

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java:
```diff
@@ -0,0 +1,173 @@
+        this.inputBufferSize = inputBufferSize;
+    }
+
+    private ByteBuf pull(ByteBufAllocator alloc) {
```
nit: static ?

slandelle replied:
Can't. It uses `decoder` which is not static.

normanmaurer commented on codec/src/test/java/io/netty/handler/codec/compression/BrotliDecoderTest.java:
```diff
@@ -0,0 +1,51 @@
+  @BeforeClass
+  public static void assumeBrotli() throws Throwable {
+//    Assume.assumeTrue("Brotli support is not available on this platform", Brotli.isAvailable());
+    Brotli.ensureAvailability();
```
this needs to be changed

slandelle replied:
please see answer above

normanmaurer commented on codec/src/test/java/io/netty/handler/codec/compression/BrotliDecoderTest.java:
```diff
@@ -0,0 +1,51 @@
+  public void initChannel() {
+    channel = new EmbeddedChannel(new BrotliDecoder());
+  }
+}
```
please use 4 spaces not 2

slandelle commented:
@normanmaurer Please hold reviews for now. I'm investigating a test failure I'm having a hard time reproducing locally. I have to stay the test suite is very weird, with  static `AbstractDecoderTest#compressedBytesLarge` actually loaded from ctor. This seems to be some JUnit hack, but this hack crashes with NPE when trying to run one single test, eg `mvn -Dtest=io.netty.handler.codec.compression.BrotliDecoderTest test`.

normanmaurer commented:
@slandelle let me know when I should take a look again

slandelle commented:
@normanmaurer I'm waiting for this build to pass (looks promising) and then I'll push some clean ups.
I gave up on extending AbstractDecoderTest whose initialization and execution lifecycle I don't trust and that was most likely causing the test failures since this morning.
Do you want me to squash the commits right now for clarity?

slandelle commented on codec/src/test/java/io/netty/handler/codec/compression/BrotliDecoderTest.java:
```diff
@@ -0,0 +1,162 @@
+      decompressed.addComponent(true, msg);
+    }
+    return decompressed;
+  }
```
done

normanmaurer commented:
@slandelle yep squashing sounds good... Also please open an issue for the "test init stuff" and explain what the problem is. Seems like we want to fix that as well

hyperxpro commented:
Finally, this PR closes #6899.

slandelle commented:
@normanmaurer squashed

slandelle commented:
CI pipeline's still not done after 3hrs???

normanmaurer commented:
@slandelle seems so... it is still "spinning" but I also don't get a console output. Maybe I need to bounce it again

hyperxpro commented:
GitHub Actions terminal is buggy. Can't see any console output.

slandelle commented:
Looks like stars are finally aligned!!!

idelpivnitskiy commented on codec/src/test/java/io/netty/handler/codec/compression/BrotliDecoderTest.java:
```diff
@@ -0,0 +1,161 @@
+import static org.junit.Assert.assertTrue;
+
+@RunWith(Theories.class)
+public class BrotliDecoderTest {
```
+1 to @normanmaurer's suggestion to create an issue, describing why `AbstractDecoderTest` could not be used with `BrotliDecoder`.

> Please hold reviews for now. I'm investigating a test failure I'm having a hard time reproducing locally. I have to stay the test suite is very weird, with  static `AbstractDecoderTest#compressedBytesLarge` actually loaded from ctor. This seems to be some JUnit hack, but this hack crashes with NPE when trying to run one single test, eg `mvn -Dtest=io.netty.handler.codec.compression.BrotliDecoderTest test`.

IIUC, it might be some visibility issue between jUnit threads

slandelle replied:
https://github.com/netty/netty/issues/11233

idelpivnitskiy reviewed (approved):
Looks great!
Thank you @slandelle for the whole effort of pushing this forward!
And @hyperxpro for the help of the way

slandelle commented:
Also, lots of credit go to @ericmolle who did an amazing work for building a toolchain for compiling brotli on CentOS6!

hyperxpro commented:
So, is this ready to merge? @normanmaurer

slandelle commented:
Awesome, thanks a lot to @idelpivnitskiy and you for the reviews!

Do you have any ETA for the release so we can plan on our side and remove the duplicated code?

normanmaurer commented:
@slandelle end of this week
